File: spring-data/src/main/java/dev/cerbos/queryplan/springdata/Scope.java
/*
 * Copyright 2021-2026 Zenauth Ltd.
 * SPDX-License-Identifier: Apache-2.0
 */

package dev.cerbos.queryplan.springdata;

import jakarta.persistence.criteria.AbstractQuery;
import jakarta.persistence.criteria.From;
import jakarta.persistence.criteria.JoinType;
import jakarta.persistence.criteria.Path;
import jakarta.persistence.metamodel.Bindable;
import jakarta.persistence.metamodel.ManagedType;
import jakarta.persistence.metamodel.PluralAttribute;
import jakarta.persistence.metamodel.SingularAttribute;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;

/**
 * Resolves a plan variable such as {@code request.resource.attr.foo} or a lambda-scoped
 * {@code t.name} to a scalar JPA {@link Path} or a relation, relative to the current
 * (sub)query's {@code From}.
 */
sealed interface Scope permits Scope.RootScope, Scope.LambdaScope {

    /**
     * Resolves {@code cerbosVar} against this scope. Never returns null: throws
     * {@link UnmappedAttributeException} for a name the mapping does not cover and
     * {@link MalformedPlanException} for an unbound lambda reference.
     *
     * <p>A dotted variable matches the longest registered Relation prefix, and the rest is
     * walked through nested {@code fields()} maps. A {@link LambdaScope} resolves its own
     * variable against the joined element and delegates everything else outward.
     */
    Resolution resolve(String cerbosVar);

    /**
     * The JPA path to compare {@code cerbosVar} as. Throws for a relation, or for a Field with
     * no column on this scope's {@code From}.
     */
    default Path<?> path(String cerbosVar) {
        if (resolve(cerbosVar) instanceof ResolvedScalar scalar) {
            if (scalar.path() == null) {
                throw Refusals.unknownAttribute(cerbosVar);
            }
            return scalar.path();
        }
        throw Refusals.unmapped(
                "Attribute " + cerbosVar + " is a Relation; cannot resolve as a scalar path");
    }

    From<?, ?> from();

    AbstractQuery<?> parentQuery();

    /** What a Cerbos plan variable denotes in a scope: a scalar value, or a relation. */
    sealed interface Resolution permits ResolvedScalar, ResolvedRelation {}

    /**
     * A variable denoting a single value, and the mapping it was resolved through.
     *
     * <p>For the bare lambda variable ({@code t} in {@code tags.exists(t, ...)}) the mapping
     * is the {@link AttributeMapping.Relation} it ranges over. {@code path} is null when the
     * Field is reachable only through a relation chain, so it has no column on this scope's
     * entity; {@link #path} reports that as an unknown attribute.
     */
    record ResolvedScalar(Path<?> path, AttributeMapping mapping) implements Resolution {}

    /**
     * A relation-valued variable: the Relations to join through, in hop order. A multi-hop
     * chain denotes the flattened union of the last hop's elements.
     *
     * <p>{@code owner} is the scope whose {@code from()} holds the first hop. A subquery over
     * the relation must correlate that {@code From}; any other could silently query a
     * same-named collection on the wrong entity.
     */
    record ResolvedRelation(Scope owner, List<AttributeMapping.Relation> chain)
            implements Resolution {
        public ResolvedRelation {
            chain = List.copyOf(chain);
        }

        boolean isChained() {
            return chain.size() > 1;
        }

        AttributeMapping.Relation tail() {
            return chain.get(chain.size() - 1);
        }
    }

    static Scope root(From<?, ?> root, AbstractQuery<?> query, Map<String, AttributeMapping> mapper) {
        return new RootScope(root, query, mapper);
    }

    static Scope lambda(From<?, ?> from, AbstractQuery<?> parentQuery,
                        AttributeMapping.Relation relation, String lambdaVar, Scope outer) {
        return new LambdaScope(from, parentQuery, relation, lambdaVar, outer);
    }

    /**
     * Re-roots the scope chain for use inside subquery {@code sub}, which correlated
     * {@code target}'s {@code from()}. The {@code target} level is re-rooted at
     * {@code correlated}; levels between {@code scope} and {@code target} keep their Froms but
     * build deeper subqueries against {@code sub}. {@code target} is matched by identity, since
     * it is the {@link ResolvedRelation#owner()} returned on this chain.
     */
    static Scope rebaseAt(Scope scope, Scope target, From<?, ?> correlated, AbstractQuery<?> sub) {
        if (scope == target) {
            if (scope instanceof RootScope rs) {
                return new RootScope(correlated, sub, rs.mapper());
            }
            LambdaScope ls = (LambdaScope) scope;
            return new LambdaScope(correlated, sub, ls.relation(), ls.lambdaVar(), ls.outer());
        }
        if (scope instanceof LambdaScope ls && ls.outer() != null) {
            return new LambdaScope(ls.from(), sub, ls.relation(), ls.lambdaVar(),
                    rebaseAt(ls.outer(), target, correlated, sub));
        }
        // An adapter invariant, not a plan refusal. ScopeTest pins the exception type.
        throw new IllegalArgumentException(
                "Relation owner scope is not on the current resolution chain");
    }

    record RootScope(From<?, ?> from, AbstractQuery<?> parentQuery, Map<String, AttributeMapping> mapper)
            implements Scope {
        @Override
        public Resolution resolve(String cerbosVar) {
            if (mapper.get(cerbosVar) instanceof AttributeMapping.Field f) {
                return new ResolvedScalar(traversePath(from, f.jpaPath()), f);
            }
            // Otherwise a registered Relation or a dotted path off one.
            RelationChain chain = resolveRelationChain(mapper, cerbosVar);
            if (chain == null) {
                throw Refusals.unknownAttribute(cerbosVar);
            }
            if (chain.tail() == null) {
                return new ResolvedRelation(this, chain.relations());
            }
            // A Field reached through the chain has no column here; see ResolvedScalar.
            return new ResolvedScalar(null, chain.tail());
        }
    }

    /**
     * Scope inside a collection lambda. The lambda variable resolves against the joined
     * element; anything else delegates to {@code outer}, the enclosing scope re-rooted inside
     * the subquery.
     */
    record LambdaScope(From<?, ?> from, AbstractQuery<?> parentQuery,
                       AttributeMapping.Relation relation, String lambdaVar,
                       Scope outer) implements Scope {

        private boolean isLambdaRef(String cerbosVar) {
            return cerbosVar.equals(lambdaVar) || cerbosVar.startsWith(lambdaVar + ".");
        }

        @Override
        public Resolution resolve(String cerbosVar) {
            if (!isLambdaRef(cerbosVar)) {
                // An outer reference is resolved and owned further out.
                if (outer != null) {
                    return outer.resolve(cerbosVar);
                }
                throw Refusals.notALambdaReference(cerbosVar, lambdaVar);
            }
            String suffix = extractLambdaSuffix(cerbosVar, lambdaVar);
            if (suffix.isEmpty()) {
                // The bare lambda variable is the element itself.
                return new ResolvedScalar(memberPath(from, relation, suffix), relation);
            }
            List<AttributeMapping.Relation> chain = relationChain(suffix);
            if (chain != null) {
                return new ResolvedRelation(this, chain);
            }
            // A member scalar: its nested mapping if registered, else the suffix as a JPA path.
            AttributeMapping nested = relation.fields().get(suffix);
            return new ResolvedScalar(memberPath(from, relation, suffix),
                    nested != null ? nested : AttributeMapping.field(suffix));
        }

        /**
         * The Relations {@code suffix} walks through the element's nested {@code fields()}, or
         * {@code null} if any part is not a Relation.
         */
        private List<AttributeMapping.Relation> relationChain(String suffix) {
            List<AttributeMapping.Relation> chain = new ArrayList<>();
            AttributeMapping.Relation current = relation;
            for (String part : suffix.split("\\.")) {
                if (!(current.fields().get(part) instanceof AttributeMapping.Relation next)) {
                    return null;
                }
                chain.add(next);
                current = next;
            }
            return chain;
        }
    }

    /** A chain of Relations, ending in the Field {@code tail}, or in the last Relation when null. */
    record RelationChain(List<AttributeMapping.Relation> relations, AttributeMapping.Field tail) {}

    /**
     * Resolves {@code cerbosVar} by its longest registered Relation prefix that the rest of the
     * path can be walked from, or returns {@code null}.
     */
    static RelationChain resolveRelationChain(Map<String, AttributeMapping> mapper, String cerbosVar) {
        AttributeMapping direct = mapper.get(cerbosVar);
        if (direct instanceof AttributeMapping.Relation rel) {
            return new RelationChain(List.of(rel), null);
        }
        String[] parts = cerbosVar.split("\\.");
        for (int i = parts.length - 1; i > 0; i--) {
            String prefix = String.join(".", Arrays.copyOfRange(parts, 0, i));
            if (mapper.get(prefix) instanceof AttributeMapping.Relation head) {
                RelationChain chain = walkSuffix(head, Arrays.copyOfRange(parts, i, parts.length));
                if (chain != null) {
                    return chain;
                }
            }
        }
        return null;
    }

    /**
     * Walks {@code suffix} through {@code head}'s nested {@code fields()}. Every part must be a
     * Relation except the last, which may be a Field; otherwise returns {@code null}.
     */
    private static RelationChain walkSuffix(AttributeMapping.Relation head, String[] suffix) {
        List<AttributeMapping.Relation> chain = new ArrayList<>(List.of(head));
        AttributeMapping.Relation current = head;
        for (int i = 0; i < suffix.length; i++) {
            AttributeMapping next = current.fields().get(suffix[i]);
            if (next instanceof AttributeMapping.Relation relation) {
                chain.add(relation);
                current = relation;
            } else if (next instanceof AttributeMapping.Field field && i == suffix.length - 1) {
                return new RelationChain(chain, field);
            } else {
                return null;
            }
        }
        return new RelationChain(chain, null);
    }

    /**
     * A member path off a join over {@code rel}. With no {@code memberField}: the
     * {@code defaultMemberField}, else the element itself. Otherwise the nested Field mapping,
     * else the raw name as a JPA path.
     */
    static Path<?> memberPath(From<?, ?> from, AttributeMapping.Relation rel, String memberField) {
        if (memberField == null || memberField.isEmpty()) {
            if (rel.defaultMemberField() != null && !rel.defaultMemberField().isEmpty()) {
                return from.get(rel.defaultMemberField());
            }
            return (Path<?>) from;
        }
        AttributeMapping nested = rel.fields().get(memberField);
        if (nested instanceof AttributeMapping.Field f) {
            return traversePath(from, f.jpaPath());
        }
        return traversePath(from, memberField);
    }

    /**
     * Walks a dotted JPA path, LEFT-joining each association before the last segment.
     *
     * <p>{@code Path.get()} would inner-join, dropping the row from the whole query when the
     * association is absent, even when another {@code OR} branch allows it. A LEFT join makes
     * the missing hop NULL, which only makes its own comparison UNKNOWN. Embeddables stay on
     * {@code get()}.
     */
    static Path<?> traversePath(From<?, ?> from, String dottedJpaPath) {
        String[] parts = dottedJpaPath.split("\\.");
        Path<?> p = from;
        for (int i = 0; i < parts.length; i++) {
            boolean last = i == parts.length - 1;
            if (!last && p instanceof From<?, ?> f && isAssociation(f, parts[i])) {
                p = f.join(parts[i], JoinType.LEFT);
            } else {
                p = p.get(parts[i]);
            }
        }
        return p;
    }

    /**
     * Whether {@code part} is an association on {@code from}'s type. Asks the metamodel, because
     * resolving the path would register an implicit inner join.
     */
    private static boolean isAssociation(From<?, ?> from, String part) {
        ManagedType<?> managed = managedTypeOf(from);
        if (managed == null) {
            return false;
        }
        try {
            return managed.getAttribute(part).isAssociation();
        } catch (IllegalArgumentException notAnAttribute) {
            // Let get() raise the provider's own error for an unknown name.
            return false;
        }
    }

    /**
     * The managed type of {@code from}. A {@code Join}'s model is the attribute it was created
     * from, so the target type is read from that attribute.
     */
    private static ManagedType<?> managedTypeOf(From<?, ?> from) {
        Bindable<?> model = from.getModel();
        if (model instanceof ManagedType<?> managed) {
            return managed;
        }
        if (model instanceof SingularAttribute<?, ?> singular
                && singular.getType() instanceof ManagedType<?> target) {
            return target;
        }
        if (model instanceof PluralAttribute<?, ?, ?> plural
                && plural.getElementType() instanceof ManagedType<?> target) {
            return target;
        }
        return null;
    }

    static String extractLambdaSuffix(String variable, String lambdaVar) {
        if (variable.equals(lambdaVar)) {
            return "";
        }
        String prefix = lambdaVar + ".";
        if (!variable.startsWith(prefix)) {
            throw Refusals.notALambdaReference(variable, lambdaVar);
        }
        return variable.substring(prefix.length());
    }
}

File: exposed/src/main/kotlin/dev/cerbos/queryplan/exposed/Scope.kt
package dev.cerbos.queryplan.exposed

import org.jetbrains.exposed.v1.core.Alias
import org.jetbrains.exposed.v1.core.Column
import org.jetbrains.exposed.v1.core.Expression
import org.jetbrains.exposed.v1.core.Table

/**
 * Where a plan variable is looked up: the root resource, or the element of an enclosing lambda.
 *
 * [resolve] is TOTAL: every variable lands in one of [Resolution]'s two arms or throws naming why.
 * There is no null return and no second classifier.
 *
 * Three rules live here and only here:
 *  - **longest-prefix chain walking** — a dotted variable is matched against the longest MAPPED
 *    prefix and the remainder walked through nested [AttributeMapping.Relation.fields];
 *  - **lambda delegation** — a [LambdaScope] resolves its own variable against the element alias
 *    and delegates everything else OUTWARD, keeping the outer scope as the owner;
 *  - **owner anchoring** — a [Resolution.Collection] carries the scope that owns its first hop,
 *    which is where a subquery over it must correlate.
 */
internal sealed interface Scope {
    fun resolve(variable: String): Resolution

    /**
     * A column of THIS scope's own table, as SQL reads it here: bare at the root, through the
     * element alias inside a lambda.
     *
     * [AttributeMapping.Relation.from] is a column on the PARENT scope's table, so a subquery over
     * a relation declared inside a lambda has to correlate against the alias rather than the bare
     * table. Reading it bare still builds and still returns rows — the wrong ones, from whichever
     * row of the outer table the planner's optimiser happens to pair it with.
     */
    fun read(column: Column<*>): Column<*>

    /** [resolve], for a position that needs one value per row. */
    fun scalar(variable: String): Resolution.Scalar = when (val resolution = resolve(variable)) {
        is Resolution.Scalar -> resolution
        is Resolution.Collection -> throw Refusals.unmapped(
            "$variable is mapped as a relation, but this position needs a scalar column",
        )
    }

    /** [resolve], for a position that needs a collection. */
    fun collection(variable: String): Resolution.Collection = when (val resolution = resolve(variable)) {
        is Resolution.Collection -> resolution
        is Resolution.Scalar -> throw Refusals.unmapped(
            "$variable is mapped as a column, but this position needs a relation",
        )
    }
}

internal sealed interface Resolution {
    /**
     * One value per row of the resolving scope.
     *
     * [expression] is what goes into SQL: the column itself, the same column read through a
     * subquery alias, or a correlated scalar subquery when the variable reaches through a to-one
     * relation (NULL when no row correlates, which is already CEL's missing-attribute error).
     *
     * [column] is the DECLARED column whatever [expression] is. It carries the type knowledge the
     * plan does not: `column.columnType` says boolean, text, numeric or temporal.
     */
    class Scalar(
        val variable: String,
        val expression: Expression<*>,
        val column: Column<*>,
        val field: AttributeMapping.Field,
    ) : Resolution

    /**
     * A collection reached through one or more relations, outermost first; the last hop's table is
     * the element table.
     *
     * [owner] is the scope that OWNS the first hop, which is not always the scope that resolved
     * the variable: `request.resource.attr.tags` inside a `categories` lambda belongs to the root.
     * Anchoring the subquery to the wrong scope still builds and still returns rows, the wrong
     * ones, so the owner travels with the resolution rather than being re-derived.
     */
    class Collection(
        val variable: String,
        val owner: Scope,
        val hops: List<AttributeMapping.Relation>,
    ) : Resolution {
        init {
            require(hops.isNotEmpty()) { "a collection resolution needs at least one hop" }
        }

        val tail: AttributeMapping.Relation get() = hops.last()

        /**
         * The hops that are merely TRAVERSED rather than iterated. Each is a to-one parent on the
         * check side — CEL cannot dot through a list — so an absent one is a missing-path error,
         * and a subquery rooted at the resource row cannot tell it apart from a childless parent.
         */
        val leadingHops: List<AttributeMapping.Relation> get() = hops.subList(0, hops.size - 1)
    }
}

/** The root resource: mapped columns are read bare, and relations correlate against its own row. */
internal class RootScope(private val translation: Translation) : Scope {
    private val resolved = HashMap<String, Resolution>()

    override fun read(column: Column<*>): Column<*> = column

    override fun resolve(variable: String): Resolution = resolved.getOrPut(variable) { classify(variable) }

    private fun classify(variable: String): Resolution {
        when (val direct = translation.options.mapping.resolve(variable)) {
            is AttributeMapping.Field -> return Resolution.Scalar(variable, read(direct.column), direct.column, direct)
            is AttributeMapping.Relation ->
                return ChainWalk(listOf(direct), null).resolveAt(translation, this, variable)
            null -> Unit
        }
        // Not mapped under its whole name, so it is a dotted suffix off a mapped relation. The
        // resolver may be a FUNCTION rather than a table, so the prefixes are probed one by one,
        // longest first: the mapping cannot be enumerated and asked which keys are prefixes.
        val parts = variable.split('.')
        for (length in parts.size - 1 downTo 1) {
            val prefix = parts.subList(0, length).joinToString(".")
            val head = translation.options.mapping.resolve(prefix) as? AttributeMapping.Relation ?: continue
            val walked = ChainWalk.from(head, parts.subList(length, parts.size)) ?: continue
            // The mapped prefix is itself the first hop: `from` walks the nested fields BEYOND the
            // relation it starts at, so dropping the head here would correlate the remainder
            // straight off the root and skip the intermediate table entirely.
            return ChainWalk(listOf(head) + walked.hops, walked.leaf).resolveAt(translation, this, variable)
        }
        throw Refusals.unknownAttribute(variable)
    }
}

/**
 * Inside a collection lambda. The lambda's own variable resolves against the element alias;
 * everything else delegates OUTWARD, which is what keeps `request.resource.attr.tags` inside a
 * `categories` lambda owned by the root and therefore correlated against the root's row.
 */
internal class LambdaScope(
    private val translation: Translation,
    private val alias: Alias<Table>,
    private val relation: AttributeMapping.Relation,
    private val lambdaVariable: String,
    private val outer: Scope,
) : Scope {
    private val resolved = HashMap<String, Resolution>()

    override fun read(column: Column<*>): Column<*> {
        if (column.table !== alias.delegate) {
            throw Refusals.internal(
                "Column ${column.table.tableName}.${column.name} is not a column of " +
                    "${alias.delegate.tableName}, which this lambda iterates",
            )
        }
        return alias[column]
    }

    override fun resolve(variable: String): Resolution = resolved.getOrPut(variable) { classify(variable) }

    private fun classify(variable: String): Resolution {
        if (variable == lambdaVariable) {
            val element = relation.element ?: throw RelationRefusals.noElementColumn(variable, relation)
            return Resolution.Scalar(variable, read(element.column), element.column, element)
        }
        val prefix = "$lambdaVariable."
        if (!variable.startsWith(prefix)) {
            return outer.resolve(variable)
        }
        val walked = ChainWalk.from(relation, variable.substring(prefix.length).split('.'))
            ?: throw Refusals.unknownAttribute(variable)
        if (walked.hops.isEmpty()) {
            val leaf = walked.leaf ?: throw Refusals.internal("A member walk ended at neither a column nor a relation")
            return Resolution.Scalar(variable, read(leaf.column), leaf.column, leaf)
        }
        return walked.resolveAt(translation, this, variable)
    }
}

/**
 * A dotted reference walked onto storage: the relation [hops] it passes through, and the [leaf]
 * column it ends at when it ends at one rather than at a relation.
 *
 * [hops] is relative to the scope that walked it — [ChainWalk.from] never includes the relation it
 * started from, and [RootScope] prepends the mapped head itself.
 */
internal class ChainWalk(
    val hops: List<AttributeMapping.Relation>,
    val leaf: AttributeMapping.Field?,
) {
    /**
     * The resolution this walk denotes, anchored at [owner].
     *
     * A path that reaches a to-many relation is a COLLECTION; one that ends at a column having
     * passed only through to-one relations is a scalar read through a correlated subquery. The
     * two are not interchangeable: an aggregate over a to-one hop and a scalar projection through
     * it differ exactly where a corpus action discriminates them.
     */
    fun resolveAt(translation: Translation, owner: Scope, variable: String): Resolution {
        if (leaf == null) {
            val tail = hops.last()
            if (tail.cardinality == AttributeMapping.Relation.Cardinality.MANY) {
                return Resolution.Collection(variable, owner, hops)
            }
            // A to-one relation named as a value denotes its element column, the same way the bare
            // lambda variable of a scalar collection does.
            val element = tail.element ?: throw RelationRefusals.noElementColumn(variable, tail)
            return scalarThroughHops(translation, owner, variable, element)
        }
        val plural = hops.firstOrNull { it.cardinality == AttributeMapping.Relation.Cardinality.MANY }
        if (plural != null) {
            throw RelationRefusals.memberOfToMany(variable, plural)
        }
        return scalarThroughHops(translation, owner, variable, leaf)
    }

    private fun scalarThroughHops(
        translation: Translation,
        owner: Scope,
        variable: String,
        field: AttributeMapping.Field,
    ): Resolution.Scalar = Resolution.Scalar(
        variable,
        translation.subqueries.scalarThroughHops(owner, hops, field.column),
        field.column,
        field,
    )

    companion object {
        /**
         * Walk [parts] through [start]'s nested fields, or `null` as soon as a segment is unmapped
         * or a non-final segment is a column. The returned hops do NOT include [start].
         */
        fun from(start: AttributeMapping.Relation, parts: List<String>): ChainWalk? {
            val hops = mutableListOf<AttributeMapping.Relation>()
            var current = start
            parts.forEachIndexed { index, part ->
                when (val next = current.fields[part]) {
                    null -> return null
                    is AttributeMapping.Relation -> {
                        hops += next
                        current = next
                    }
                    is AttributeMapping.Field ->
                        return if (index == parts.lastIndex) ChainWalk(hops, next) else null
                }
            }
            return ChainWalk(hops, null)
        }
    }
}

/**
 * The column standing for one member of [relation]'s elements: the named one, or the element
 * column itself when the member is unnamed (`t` rather than `t.name`).
 *
 * A member the mapping does not declare is refused rather than guessed at a column name: guessing
 * is how a subquery ends up comparing the wrong column and returning rows the PDP denies.
 */
internal fun relationMemberColumn(
    relation: AttributeMapping.Relation,
    member: String,
    reference: String,
): Column<*> {
    if (member.isEmpty()) {
        return (relation.element ?: throw RelationRefusals.noElementColumn(reference, relation)).column
    }
    return when (val mapped = relation.fields[member]) {
        is AttributeMapping.Field -> mapped.column
        is AttributeMapping.Relation -> throw Refusals.unmapped(
            "$reference names ${relation.table.tableName}'s '$member' relation, but this position needs a column",
        )
        null -> throw Refusals.unknownAttribute(reference)
    }
}
